packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -358,6 +323,46 @@ export function createVM<HostNode, HostElement>(
     return vm;
 }
 
+const transitivelyNativeVMs = new WeakSet<VM>();
+
+function computeShadowMode(vm: VM) {
+    const { def, renderer } = vm;
+    const { isNativeShadowDefined, isSyntheticShadowDefined } = renderer;
+
+    let shadowMode;
+    if (isSyntheticShadowDefined) {
+        if (def.renderMode === RenderMode.Light) {
+            // ShadowMode.Native implies "not synthetic shadow" which is consistent with how
+            // everything defaults to native when the synthetic shadow polyfill is unavailable.
+            shadowMode = ShadowMode.Native;
+        } else if (isNativeShadowDefined) {
+            if (def.shadowSupportMode === ShadowSupportMode.Any) {
+                shadowMode = ShadowMode.Native;
+                transitivelyNativeVMs.add(vm);
+            } else {
+                const shadowAncestor = getNearestShadowAncestor(vm);
+                if (!isNull(shadowAncestor) && transitivelyNativeVMs.has(shadowAncestor)) {
```

Comment:
Is the `WeakSet` the best way to store such information?

My mental model is the following. Weaksets are great for objects to store objects you don't own: you don't mutate the object and you don't prevent the object to be garbage collected. In this case, since we own the `VM` object and since it's an internal object, I think a better approach would be to store this information directly on the `VM`.

Reply:
Good point. We're should just check the `shadowMode`, no need for a WeakSet.